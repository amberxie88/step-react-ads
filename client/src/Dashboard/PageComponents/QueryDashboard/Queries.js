/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import axios from 'axios';
import * as HttpStatus from 'http-status-codes';
import { makeStyles } from '@material-ui/core/styles';
import QueryResults from './QueryResults';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import Title from '../../Utilities/Title';

const useStyles = makeStyles((theme) => ({
  root: {
    margin: theme.spacing(1),
  },
}));

//JSON data parsing here
function addKey(index, responseObj) {
  responseObj.id = index;
  return responseObj;
}

function parseRows(response) {
  const responseRows = [];
  let i;
  for (i = 0; i < response.length; i++) {
    responseRows.push(addKey(i, response[i]));
  }
  return responseRows;
}

class Query extends React.Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.handleQuery = this.handleQuery.bind(this);
    this.state = {
      value: '',
      rows: [],
      fields: [],
      rowsPerPage: 5,
      status: 'inputRequired',
      errorMessage: '',
    };
  }

  async handleQuery() {
    const query = this.state.value;
    this.setState({ status: 'loading' });
    try {
      const { data } = await axios.post(
        '/campaign',
        new URLSearchParams({ query }),
      );
      if (data.meta.status !== HttpStatus.OK.toString()) {
        throw new Error(data.meta.message);
      } else {
        this.setState({
          rows: parseRows(data.response),
          fields: data.fieldmask,
          status: 'loaded',
        });
      }
    } catch (err) {
      console.log(err.message);
      this.setState({ status: 'error', errorMessage: err.message });
    }
  }

  pickContentToDisplay() {
    switch (this.state.status) {
      case 'inputRequired':
        return <Title> Enter a query to see the results </Title>;
      case 'loading':
        return <Title> Loading ... </Title>;
      case 'loaded':
        return (
          <QueryResults
            rows={this.state.rows}
            fields={this.state.fields}
            rowsPerPage={this.state.rowsPerPage}
          />
        );
      case 'error':
        return (
          <Title>
            {"Something Went Wrong. Here's the Error Message: " +
              this.state.errorMessage}
          </Title>
        );
      default:
        return <Title> Something Went Wrong</Title>;
    }
  }

  handleChange(event) {
    this.setState({ value: event.target.value });
  }

  render() {
    // const rows = this.state.rows;
    // const fields = this.state.fields;
    return (
      <React.Fragment>
        <Title>Query Here</Title>
        <TextField
          value={this.state.value}
          onChange={this.handleChange}
          id="standard-full-width"
          label="Enter Query"
          style={{ margin: 8 }}
          placeholder="Placeholder"
          helperText="make sure to select an account!"
          fullWidth
          margin="normal"
          InputLabelProps={{
            shrink: true,
          }}
        />
        <SubmitButton onClick={this.handleQuery} />
        {this.pickContentToDisplay()}
      </React.Fragment>
    );
  }
}

function SubmitButton(props) {
  const classes = useStyles();
  return (
    <div className={classes.root}>
      <Button variant="outlined" onClick={props.onClick}>
        Submit
      </Button>
    </div>
  );
}

export default Query;
