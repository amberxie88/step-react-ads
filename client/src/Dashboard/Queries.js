import React from 'react';
import { makeStyles } from '@material-ui/core/styles';

import TextField from '@material-ui/core/TextField';

export default function Query() {
  return (
    <TextField
      id="standard-full-width"
      label="Enter Query"
      style={{ margin: 8 }}
      placeholder="Placeholder"
      helperText="helper text"
      fullWidth
      margin="normal"
      InputLabelProps={{
        shrink: true,
      }}
    />
  );
}
