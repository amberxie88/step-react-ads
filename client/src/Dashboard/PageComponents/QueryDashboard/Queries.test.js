import React from 'react';
import Queries from './Queries';
React.useLayoutEffect = React.useEffect; // Necessary to supress warning during tests. The tests use server side rendering which doesn't support useLayoutEffect.
import { configure, shallow } from 'enzyme';
import { waitForState } from 'enzyme-async-helpers';
import Adapter from 'enzyme-adapter-react-16';
import axios from 'axios';

configure({ adapter: new Adapter() });
jest.mock('axios');

describe('Queries Unit Testing', () => {
  it('Queries correctly displays data with properly formatted input', async () => {
    const mockedAPICall = {
      data: {
        response: [
          {
            'campaign.id': '10375582075',
            'campaign.name': 'Sales-Search-1',
            'metrics.clicks': '0',
          },
        ],
        fieldmask: ['campaign.id', 'campaign.name', 'metrics.clicks'],
        meta: { status: '200' },
      },
    };
    axios.post.mockImplementationOnce(() => Promise.resolve(mockedAPICall));
    const component = shallow(<Queries />);
    component.find('SubmitButton').simulate('click'); //click the submit button. This should trigger the api call
    await waitForState(component, (state) => state.rows !== []); //wait for rows state variable to change from initial state
    expect(component.state('status')).toEqual('loaded'); //TODO change to use actual code instead
    expect(component.state('rows')).toEqual([
      {
        'campaign.id': '10375582075',
        'campaign.name': 'Sales-Search-1',
        id: 0,
        'metrics.clicks': '0',
      },
    ]);
    expect(component.state('fields')).toEqual([
      'campaign.id',
      'campaign.name',
      'metrics.clicks',
    ]);
    const QueryResultsHTML = component.find('QueryResults').html();
    expect(QueryResultsHTML).toMatchSnapshot(); //displayed state also needs to match expected results
  });

  it('Queries correctly displays error with incorrectly formatted input', async () => {
    const mockedAPICall = {
      data: {
        meta: { status: '400', message: 'Some Kind of Error' },
      },
    };
    axios.post.mockImplementationOnce(() => Promise.resolve(mockedAPICall));
    const component = shallow(<Queries />);
    component.find('SubmitButton').simulate('click'); //click the submit button. This should trigger the api call
    await waitForState(component, (state) => state.rows !== []); //wait for rows state variable to change from initial state
    expect(component.state('status')).toEqual('error');
    expect(component.state('errorMessage')).toEqual('Some Kind of Error');
    const ErrorHTML = component.find('Title').at(1).html();
    expect(ErrorHTML).toMatchSnapshot(); //displayed state also needs to match expected results
  });
});
