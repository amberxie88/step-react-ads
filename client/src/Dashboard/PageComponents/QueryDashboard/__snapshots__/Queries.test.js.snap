// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries Unit Testing Queries correctly displays data with properly formatted input 1`] = `"<h2 class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom\\">Query Results</h2><div><table class=\\"MuiTable-root\\"><thead class=\\"MuiTableHead-root\\"><tr class=\\"MuiTableRow-root MuiTableRow-head\\"><th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall\\" scope=\\"col\\">campaign.id</th><th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall\\" scope=\\"col\\">campaign.name</th><th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall\\" scope=\\"col\\">metrics.clicks</th></tr></thead><tbody class=\\"MuiTableBody-root\\"><tr class=\\"MuiTableRow-root\\"><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">10375582075</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">Sales-Search-1</td><td class=\\"MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall\\">0</td></tr></tbody></table><div class=\\"MuiTablePagination-root\\"><div class=\\"MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters\\"><div class=\\"MuiTablePagination-spacer\\"></div><p class=\\"MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit\\">Rows per page:</p><div class=\\"MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot\\"><div class=\\"MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input\\" tabindex=\\"0\\" role=\\"button\\" aria-haspopup=\\"listbox\\">10</div><input value=\\"10\\" aria-hidden=\\"true\\" tabindex=\\"-1\\" class=\\"MuiSelect-nativeInput\\"/><svg class=\\"MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M7 10l5 5 5-5z\\"></path></svg></div><p class=\\"MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit\\">1-1 of 1</p><div class=\\"MuiTablePagination-actions\\"><button class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled\\" tabindex=\\"-1\\" type=\\"button\\" disabled=\\"\\" title=\\"Previous page\\" aria-label=\\"Previous page\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z\\"></path></svg></span></button><button class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled\\" tabindex=\\"-1\\" type=\\"button\\" disabled=\\"\\" title=\\"Next page\\" aria-label=\\"Next page\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z\\"></path></svg></span></button></div></div></div></div>"`;

exports[`Queries Unit Testing Queries correctly displays error with incorrectly formatted input 1`] = `"<h2 class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom\\">Something Went Wrong. Here&#x27;s the Error Message: Some Kind of Error</h2>"`;

exports[`Queries Unit Testing Queries correctly displays unexpected error 1`] = `"<h2 class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom\\">Something Went Wrong. Here&#x27;s the Error Message: Cannot read property &#x27;status&#x27; of undefined</h2>"`;
